engine/access/ingestion2/finalized_block_processor.go

```diff
@@ -0,0 +1,166 @@
+func NewFinalizedBlockProcessor(
+	log zerolog.Logger,
+	state protocol.State,
+	blocks storage.Blocks,
+	executionResults storage.ExecutionResults,
+	finalizedProcessedHeight storage.ConsumerProgressInitializer,
+	syncer *CollectionSyncer,
+	collectionExecutedMetric module.CollectionExecutedMetric,
+) (*FinalizedBlockProcessor, error) {
+	reader := jobqueue.NewFinalizedBlockReader(state, blocks)
+	finalizedBlock, err := state.Final().Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not get finalized block header: %w", err)
+	}
+
+	consumerNotifier := engine.NewNotifier()
+	processor := &FinalizedBlockProcessor{
+		log:                      log,
+		blocks:                   blocks,
+		executionResults:         executionResults,
+		consumerNotifier:         consumerNotifier,
+		collectionSyncer:         syncer,
+		collectionExecutedMetric: collectionExecutedMetric,
+	}
+
+	processor.consumer, err = jobqueue.NewComponentConsumer(
+		log.With().Str("module", "ingestion_block_consumer").Logger(),
+		consumerNotifier.Channel(),
+		finalizedProcessedHeight,
+		reader,
+		finalizedBlock.Height,
+		processor.processFinalizedBlockJobCallback,
+		processFinalizedBlocksWorkersCount,
+		searchAhead,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error creating finalized block jobqueue: %w", err)
+	}
+
+	return processor, nil
+}
+
+// Notify notifies the processor that a new finalized block is available for processing.
+func (p *FinalizedBlockProcessor) Notify(_ *model.Block) {
+	p.consumerNotifier.Notify()
+}
+
+// StartProcessing begins processing of finalized blocks and signals readiness when initialization is complete.
+func (p *FinalizedBlockProcessor) StartProcessing(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	p.consumer.Start(ctx)
+
+	err := util.WaitClosed(ctx, p.consumer.Ready())
+	if err == nil {
+		ready()
+	}
+
+	<-p.consumer.Done()
+}
+
+// processFinalizedBlockJobCallback is a jobqueue callback that processes a finalized block job.
+func (p *FinalizedBlockProcessor) processFinalizedBlockJobCallback(
+	ctx irrecoverable.SignalerContext,
+	job module.Job,
+	done func(),
+) {
+	block, err := jobqueue.JobToBlock(job)
+	if err != nil {
+		ctx.Throw(fmt.Errorf("failed to convert job to block: %w", err))
+	}
+
+	err = p.indexFinalizedBlock(block)
+	if err == nil {
+		done()
+		return
+	}
+
+	p.log.Error().Err(err).Str("job_id", string(job.ID())).Msg("error during finalized block processing job")
+}
+
+// indexFinalizedBlock indexes the given finalized block’s collection guarantees and execution results,
+// and requests related collections from the syncer.
+//
+// No errors are expected during normal operations.
+func (p *FinalizedBlockProcessor) indexFinalizedBlock(block *flow.Block) error {
+	// FIX: we can't index guarantees here, as we might have more than one block
+	// with the same collection as long as it is not finalized
+
+	// TODO: substitute an indexer module as layer between engine and storage
+
```

Comment:
The `FIX` isn't correct anymore since the logic is indexing the guarantees.
The `TODO` is being fixed by this PR!
```suggestion
```